Contribute visual examples of VimTeX's motions and text-editing mappings

ejmastnak opened the pull request:
This PR is motivated by [#1946](https://github.com/lervag/vimtex/issues/1946) and adds GIFs demonstrating most of VimTeX's motions and delete/change/toggle key mappings, i.e. most of mappings in `:help vimtex-mappings`.

Summary of changes:
- Created a dedicated `VISUALS.md` file (currently in the project root; could also be moved to `doc/`) holding the aforementioned GIFs and concise accompanying descriptions.
- Changed the "Screenshots" section of `README.md` to "Screenshots and GIFs" (TOC updated accordingly) and linked to `VISUALS.md` from the new  "Screenshots and GIFs" section of the README.
- Added link in the "Quick Start" section of `README.md` to my third-party
  article [Getting started with the VimTeX plugin](https://ejmastnak.github.io/tutorials/vim-latex/vimtex.html), as per our email exchange, [`@lervag`](https://github.com/lervag/).
- Unrelated, but minor update to `README.md` changing "All features are enabled by default" to "Most features are enabled by default" and noting the disabled folding and formatting features (see [c8e441a](https://github.com/lervag/vimtex/commit/c8e441a4153f8274d84848ce46be8fe96ea0f8e8)).

Notes: 
- The GIFs in `VISUALS.md` are currently stored in [github.com/ejmastnak/vimtex-media](https://github.com/ejmastnak/vimtex-media)---we will probably want to move these to [github.com/lervag/vimtex-media](https://github.com/lervag/vimtex-media), but please first see the comments at the end of [2c57a48](https://github.com/lervag/vimtex/commit/2c57a480dd7967f8f8be0628534a26eb37f58a6c).

- [`@lervag`](https://github.com/lervag/), we had discussed adding a section on third-party guides to the main README.
  I looked into content I could find online, but didn't find anything particularly captivating (mostly outdated summaries of setting up inverse search before the days of `VimtexInverseSearch` or nice but not terribly comprehensive summaries of standard commands/mappings which I'm not sure add much beyond what the main docs already offer).
  Although you have endorsed it, personally adding only [my own series](https://ejmastnak.github.io/tutorials/vim-latex/intro.html) to a third-party section feels too much like self-promotion, so I defer to you on this one.

ejmastnak commented:
I'm realizing I was a bit noisy with the large number of commits here and we will probably want to rebase this, but I leave that to the discretion of the higher authorities.

lervag commented on README.md:
```diff
@@ -142,12 +146,17 @@ here](https://github.com/lervag/vimtex/issues/1946#issuecomment-843674951).
+but if desired one can disable default mappings to define custom mappings. 
+Nearly all features are enabled by default, but each feature may be disabled if
+desired. The two exceptions are code folding and formating, which are disabled
+by default and must be manually enabled.
```
Ah, nice catch!

lervag replied:
Ah, nice catch!

lervag commented on VISUALS.md:
```diff
@@ -0,0 +1,183 @@
+# VimTeX Visualized
```
This is excellent! I like the title :)

One thing, though: I don't feel quite "up to the task" of maintaining this particular file. In case there are changes  or new features that someone wants "visualized", then I would need community help to update. Perhaps that should be clarified here?

ejmastnak replied:
Agreed. If you have an idea about phrasing this disclaimer/notice-to-the-community, please go ahead. I am also happy to take responsibility for helping maintain this file, just not sure how that would work in practice.

lervag commented on VISUALS.md:
```diff
@@ -0,0 +1,183 @@
+# VimTeX Visualized
```
This is excellent! I like the title :)

One thing, though: I don't feel quite "up to the task" of maintaining this particular file. In case there are changes  or new features that someone wants "visualized", then I would need community help to update. Perhaps that should be clarified here?

lervag commented:
> This PR is motivated by [#1946](https://github.com/lervag/vimtex/issues/1946) and adds GIFs demonstrating most of VimTeX's motions and delete/change/toggle key mappings, i.e. most of mappings in `:help vimtex-mappings`.
> 
> Summary of changes: …

Thanks! I believe this is a very good improvement of the informal documentation! I've added a few comments to the files, nothing big.

Apart from that, I hope you don't mind if I adjust the wording after merging. In particular, I want to update the quick start part to make it more "logically" structured. But I think it is better that I do it afterwards, then you could read it and comment (if desired).

> * Added link in the "Quick Start" section of `README.md` to my third-party
>       article [Getting started with the VimTeX plugin](https://ejmastnak.github.io/tutorials/vim-latex/vimtex.html), as per our email exchange, [`@lervag`](https://github.com/lervag/).

Very good! I would also recommend that you publish the article (or series) elsewhere. Your content is good and deserves an audience!

> * Unrelated, but minor update to `README.md` changing "All features are enabled by default" to "Most features are enabled by default" and noting the disabled folding and formatting features (see [c8e441a](https://github.com/lervag/vimtex/commit/c8e441a4153f8274d84848ce46be8fe96ea0f8e8)).

Thanks for noticing and fixing that!

> * The GIFs in `VISUALS.md` are currently stored in [github.com/ejmastnak/vimtex-media](https://github.com/ejmastnak/vimtex-media)---we will probably want to move these to [github.com/lervag/vimtex-media](https://github.com/lervag/vimtex-media), but please first see the comments at the end of [2c57a48](https://github.com/lervag/vimtex/commit/2c57a480dd7967f8f8be0628534a26eb37f58a6c).

Ah, yes. Licensing. To be honest, I don't care so much about the licensing. That is, in the sense that I don't mind what anyone does with whatever I share. That's why I've used the MIT license.

However, I respect other peoples concern for licensing, and I don't want to force the MIT license here. The point here is to conserve the rights of the third-party content creators. I would therefore not mind changing to the stricter CC-BY-NC-SA for `vimtex-media`. I believe we can just do this, but I'm not a lawyer and I don't know these things well. Do we need to contact @DustyTopology for his permission first?

> * [`@lervag`](https://github.com/lervag/), we had discussed adding a section on third-party guides to the main README.
>       I looked into content I could find online, but didn't find anything particularly captivating (mostly outdated summaries of setting up inverse search before the days of `VimtexInverseSearch` or nice but not terribly comprehensive summaries of standard commands/mappings which I'm not sure add much beyond what the main docs already offer).
>       Although you have endorsed it, personally adding only [my own series](https://ejmastnak.github.io/tutorials/vim-latex/intro.html) to a third-party section feels too much like self-promotion, so I defer to you on this one.

Yes, I can see this is tricky. And I agree that there are not really that much content available that is suitable for a link here. So I think it is perfectly fine that we now only add the link to your guide.

> I'm realizing I was a bit noisy with the large number of commits here and we will probably want to rebase this, but I leave that to the discretion of the higher authorities.

No worries. I don't really mind the number of commits. But you can of course feel free to clean up before merge if you want to.

lervag commented:
> This PR is motivated by [#1946](https://github.com/lervag/vimtex/issues/1946) and adds GIFs demonstrating most of VimTeX's motions and delete/change/toggle key mappings, i.e. most of mappings in `:help vimtex-mappings`.
> 
> Summary of changes: …

Thanks! I believe this is a very good improvement of the informal documentation! I've added a few comments to the files, nothing big.

Apart from that, I hope you don't mind if I adjust the wording after merging. In particular, I want to update the quick start part to make it more "logically" structured. But I think it is better that I do it afterwards, then you could read it and comment (if desired).

> * Added link in the "Quick Start" section of `README.md` to my third-party
>       article [Getting started with the VimTeX plugin](https://ejmastnak.github.io/tutorials/vim-latex/vimtex.html), as per our email exchange, [`@lervag`](https://github.com/lervag/).

Very good! I would also recommend that you publish the article (or series) elsewhere. Your content is good and deserves an audience!

> * Unrelated, but minor update to `README.md` changing "All features are enabled by default" to "Most features are enabled by default" and noting the disabled folding and formatting features (see [c8e441a](https://github.com/lervag/vimtex/commit/c8e441a4153f8274d84848ce46be8fe96ea0f8e8)).

Thanks for noticing and fixing that!

> * The GIFs in `VISUALS.md` are currently stored in [github.com/ejmastnak/vimtex-media](https://github.com/ejmastnak/vimtex-media)---we will probably want to move these to [github.com/lervag/vimtex-media](https://github.com/lervag/vimtex-media), but please first see the comments at the end of [2c57a48](https://github.com/lervag/vimtex/commit/2c57a480dd7967f8f8be0628534a26eb37f58a6c).

Ah, yes. Licensing. To be honest, I don't care so much about the licensing. That is, in the sense that I don't mind what anyone does with whatever I share. That's why I've used the MIT license.

However, I respect other peoples concern for licensing, and I don't want to force the MIT license here. The point here is to conserve the rights of the third-party content creators. I would therefore not mind changing to the stricter CC-BY-NC-SA for `vimtex-media`. I believe we can just do this, but I'm not a lawyer and I don't know these things well. Do we need to contact @DustyTopology for his permission first?

> * [`@lervag`](https://github.com/lervag/), we had discussed adding a section on third-party guides to the main README.
>       I looked into content I could find online, but didn't find anything particularly captivating (mostly outdated summaries of setting up inverse search before the days of `VimtexInverseSearch` or nice but not terribly comprehensive summaries of standard commands/mappings which I'm not sure add much beyond what the main docs already offer).
>       Although you have endorsed it, personally adding only [my own series](https://ejmastnak.github.io/tutorials/vim-latex/intro.html) to a third-party section feels too much like self-promotion, so I defer to you on this one.

Yes, I can see this is tricky. And I agree that there are not really that much content available that is suitable for a link here. So I think it is perfectly fine that we now only add the link to your guide.

> I'm realizing I was a bit noisy with the large number of commits here and we will probably want to rebase this, but I leave that to the discretion of the higher authorities.

No worries. I don't really mind the number of commits. But you can of course feel free to clean up before merge if you want to.

ejmastnak commented on README.md:
```diff
@@ -111,11 +111,15 @@ let maplocalleader = ","
+An article covering VimTeX's core features, containing plenty of examples and
+high-resolution animations intended to help new users ease into working with the
+plugin, is available at the third-party link [Getting started with the VimTeX
+plugin](https://ejmastnak.github.io/tutorials/vim-latex/vimtex.html).
```
I'm all for it, thanks!

ejmastnak commented:
> Apart from that, I hope you don't mind if I adjust the wording after merging. In particular, I want to update the quick start part to make it more "logically" structured. But I think it is better that I do it afterwards, then you could read it and comment (if desired).

Please go ahead, I have no reservations about this.

> Yes, I can see this is tricky. And I agree that there are not really that much content available that is suitable for a link here. So I think it is perfectly fine that we now only add the link to your guide.

Sounds good. I'll add a short introduction of my series to the README as you suggested in our email exchange, and then you can reword it or rearrange its position if needed, similar to how you will tweak the Quick Start section.

> However, I respect other peoples concern for licensing, and I don't want to force the MIT license here. The point here is to conserve the rights of the third-party content creators. I would therefore not mind changing to the stricter CC-BY-NC-SA for vimtex-media. I believe we can just do this, but I'm not a lawyer and I don't know these things well.

If you are willing to change vimtex-media to a CC license that would be great---I do feel a CC would be more appropriate for media. I decided to change ejmastnak/vimtex-media to the (slightly) more permissive BY-NC, which at least permits sub-licensing remixed works, and in fact wouldn't mind lowering this further to the very permissive CC-BY, which is roughly a CC equivalent of MIT (allowing essentially arbitrary use as long as copyright/attribution is given). I lean towards non-commercial since I support free sharing of content, but I recognize forcing this on a few GIFs may just be idealism. Let me know if you would prefer CC-BY. I don't mind too much one way or the other here, my only preference could be a CC license for non-software content.

> Do we need to contact @DustyTopology for his permission first?

@DustyTopology's GitHub account seems to be deleted, so I don't know if it is possible to contact him. I imagine switching from MIT to CC-BY should be fine. If we instead stick with BY-NC, which revokes MIT's commercial use clause, one possible solution is to use two directories, one with each license (although this is less clean and risks future clutter).

ejmastnak commented:
I have pushed a few more commits that should resolve the following

[On maintaining VISUALS.md]
> One thing, though: I don't feel quite "up to the task" of maintaining this particular file. In case there are changes or new features that someone wants "visualized", then I would need community help to update. Perhaps that should be clarified here?

Addressed in [58e704a](https://github.com/lervag/vimtex/pull/2362/commits/58e704a7eabf35f43fac0462d442fcb83f7f2328)

[On third-party guides]
> Yes, I can see this is tricky. And I agree that there are not really that much content available that is suitable for a link here. So I think it is perfectly fine that we now only add the link to your guide.

Addressed in [6429f84](https://github.com/lervag/vimtex/pull/2362/commits/6429f84ce4251ebd1850d08e03e575b78404b0b6)

In both cases, consider the commits a first draft, just to put ideas on the table, and feel free to edit as you see fit.

lervag commented:
> > Yes, I can see this is tricky. And I agree that there are not really that much content available that is suitable for a link here. So I think it is perfectly fine that we now only add the link to your guide.
> 
> Sounds good. I'll add a short introduction of my series to the README as you suggested in our email exchange, and then you can reword it or rearrange its position if needed, similar to how you will tweak the Quick Start section.

Great!

> > However, I respect other peoples concern for licensing, and I don't want to force the MIT license here. The point here is to conserve the rights of the third-party content creators. I would therefore not mind changing to the stricter CC-BY-NC-SA for vimtex-media. I believe we can just do this, but I'm not a lawyer and I don't know these things well.
> 
> If you are willing to change vimtex-media to a CC license that would be great---I do feel a CC would be more appropriate for media. I decided to change ejmastnak/vimtex-media to the (slightly) more permissive BY-NC, which at least permits sub-licensing remixed works, and in fact wouldn't mind lowering this further to the very permissive CC-BY, which is roughly a CC equivalent of MIT (allowing essentially arbitrary use as long as copyright/attribution is given). I lean towards non-commercial since I support free sharing of content, but I recognize forcing this on a few GIFs may just be idealism. Let me know if you would prefer CC-BY. I don't mind too much one way or the other here, my only preference could be a CC license for non-software content.

Yes, agreed. Let's move to CC BY-NC. If anyone has commercial interests, they could always contact owners directly.

> > Do we need to contact @DustyTopology for his permission first?
> 
> @DustyTopology's GitHub account seems to be deleted, so I don't know if it is possible to contact him. I imagine switching from MIT to CC-BY should be fine. If we instead stick with BY-NC, which revokes MIT's commercial use clause, one possible solution is to use two directories, one with each license (although this is less clean and risks future clutter).

You're right, his account is deleted, and I do not have any of his contact details. The repository is mine, this discussion is public, so I believe there should be no problem in just changing the license. I propose that you make the change of license a part of your future PR to vimtex-media?

----

Regarding your recent updates: It looks good. I'll merge when I have the time and then make some adjustments. These will most likely all be minor.

One minor comment, though: I really don't like when a branch pulls in master in a merge. Instead, I much prefer to have branches rebased on top of master before I merge. That way, it is _much_ easier to inspect the changes brought by any given branch.

I can manually fix this myself when I merge, unless you want to fix it yourself. If so, you would have to do a `git push --force` after you fix. Some people find it complicated to work with rebasing and such, so I don't mind doing this myself.

lervag commented:
Sorry for the late reply. Started a new job this week and have slightly too much on my mind. :p

> Ah, beginner mistake, apologies. Thank you for pointing this out, this is actually my first contribution to an open source project, and I will know for next time.

No problem! Some people don't mind this at all, I may be somewhat strict in this regard. I don't know. But I guess it's one of the benefits of being the main developer of something :)

> Given my current lack of experience, I think it best that you fix this yourself this time. In the meantime, I will read up on rebasing. I imagine [the Git Book](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) would be the standard place to start, but if you know of good references for this workflow I will definitely read through them!

Sounds good! Here are some good Git references that may be useful/of help. Note that there may possibly be better guides available today that I'm not aware of. However, I believe they should still be relevant and good.

* [Rebase without tears](https://matthew-brett.github.io/pydagogue/rebase_without_tears.html)
* [Git rebase in depth](https://git-rebase.io)
* [Comparing Workflows](https://www.atlassian.com/git/tutorials/comparing-workflows/) - this discusses various workflows and may be useful to get a nice overview of different ideas
* [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) - this is not about rebasing, but it is relevant because I try to follow this convention in VimTeX.
* [Git Koans](http://stevelosh.com/blog/2013/04/git-koans/) - this is just a little bit fun :)

> * Opened PR [#2](https://github.com/lervag/vimtex-media/pull/2) in lervag/vimtex-media to move GIF files from ejmastnak/vimtex-media to lervag/vimtex-media

Great, thanks!

> * Perhaps prematurely, perhaps conveniently, went ahead and updated the GIF source URLs in VimTeX's new VISUALS.md file to point to their new, soon-to-be home in lervag/vimtex-media. (So the GIFs in VISUALS.md will be temporarily invisible until PR [#2](https://github.com/lervag/vimtex-media/pull/2) is merged into lervag/vimtex-media.

Sounds good! We'll merge vimtex-media first.

lervag commented:
So, to summarize my actions here:

* [ ] Review and merge PR2 in vimtex-media
* [ ] Rebase then merge this PR afterwards

Please let me know if it seems I missed something!

ejmastnak commented:
> Sorry for the late reply. Started a new job this week and have slightly too much on my mind. :p

No problem at all, good luck with the transition!

> Sounds good! Here are some good Git references that may be useful/of help. Note that there may possibly be better guides available today that I'm not aware of. However, I believe they should still be relevant and good.

Thank you! I appreciate the effort you made to put together a thorough list of references.

ejmastnak commented:
> So, to summarize my actions here:
> 
>     * [x]  Review and merge PR2 in vimtex-media
> 
>     * [x]  Update license of vimtex-media
> 
>     * [ ]  Rebase then merge this PR afterwards
> 
> 
> Please let me know if it seems I missed something!

I took another look through vimtex-media and my fork of vimtex, and everything looks good. Please go ahead!

lervag commented:
> > Sorry for the late reply. Started a new job this week and have slightly too much on my mind. :p
> 
> No problem at all, good luck with the transition!

Thanks! :)

> > Sounds good! Here are some good Git references that may be useful/of help. Note that there may possibly be better guides available today that I'm not aware of. However, I believe they should still be relevant and good.
> 
> Thank you! I appreciate the effort you made to put together a thorough list of references.

No problem. Glad to help!

And finally, thanks for the very useful contribution to the project!

lervag commented:
FYI: I made a few minor adjustments to the text. The most noticable was to move your new paragraph under quick start below the video. The reasoning was that I feel it makes more sense to promote the video since it is more "quick", whereas your article does take some more time for anyone to read and study. So, we now first show a short video intro, then encourage your excellent guide, then finally stresses that people should read the docs.

Let me know if you think any of my changes are regressions!

ejmastnak commented:
> Ok, I've done the merge. I did these steps, in case you're curious:
> The final commit graph thus looks like this:

Thanks for sharing this! Doesn't look too difficult at all---I'll be ready to do it myself next time if need be.

> FYI: I made a few minor adjustments to the text. The most noticable was to move your new paragraph under quick start below the video. The reasoning was that I feel it makes more sense to promote the video since it is more "quick", whereas your article does take some more time for anyone to read and study. So, we now first show a short video intro, then encourage your excellent guide, then finally stresses that people should read the docs.

Yes, the progression from the "quick" video to the somewhat longer guide to the official docs, which are the most comprehensive, makes perfect sense to me. So I fully agree with all changes in [882641b](https://github.com/lervag/vimtex/commit/882641b16b80fde9862902f57aeb86767faad980).

> And finally, thanks for the very useful contribution to the project!

You're very welcome, I'm glad it could help. Thank *you* for the excellent plugin!
